// use-mutative@1.3.0/dist/index.js
import { create, } from 'mutative';
import { useState, useReducer, useCallback, useMemo, useRef, useEffect, } from 'react';
/**
 * `useMutative` is a hook that is similar to `useState` but it uses `mutative` to handle the state updates.
 *
 *  @example
 *
 * ```ts
 * import { act, renderHook } from '@testing-library/react';
 *
 * import { useMutative } from '../src/index';
 *
 * const { result } = renderHook(() => useMutative({ items: [1] }));
 * const [state, updateState] = result.current;
 * act(() =>
 *   updateState((draft) => {
 *     draft.items.push(2);
 *   })
 * );
 * const [nextState] = result.current;
 * expect(nextState).toEqual({ items: [1, 2] });
 * ```
 */
function useMutative(
/**
 * The initial state. You may optionally provide an initializer function to calculate the initial state.
 */
initialValue, 
/**
 * Options for the `useMutative` hook.
 */
options) {
    const patchesRef = useRef({
        patches: [],
        inversePatches: [],
    });
    //#region support strict mode and concurrent features
    const count = useRef(0);
    const renderCount = useRef(0);
    let currentCount = count.current;
    useEffect(() => {
        count.current = currentCount;
        renderCount.current = currentCount;
    });
    currentCount += 1;
    renderCount.current += 1;
    //#endregion
    const [state, setState] = useState(() => typeof initialValue === 'function' ? initialValue() : initialValue);
    const updateState = useCallback((updater) => {
        setState((latest) => {
            const updaterFn = typeof updater === 'function' ? updater : () => updater;
            const result = create(latest, updaterFn, options);
            if (options === null || options === void 0 ? void 0 : options.enablePatches) {
                // check render count, support strict mode and concurrent features
                if (renderCount.current === count.current ||
                    renderCount.current === count.current + 1) {
                    Array.prototype.push.apply(patchesRef.current.patches, result[1]);
                    // `inversePatches` should be in reverse order when multiple setState() executions
                    Array.prototype.unshift.apply(patchesRef.current.inversePatches, result[2]);
                }
                return result[0];
            }
            return result;
        });
    }, []);
    useEffect(() => {
        if (options === null || options === void 0 ? void 0 : options.enablePatches) {
            // Reset `patchesRef` when the component is rendered each time
            patchesRef.current.patches = [];
            patchesRef.current.inversePatches = [];
        }
    });
    return ((options === null || options === void 0 ? void 0 : options.enablePatches)
        ? [
            state,
            updateState,
            patchesRef.current.patches,
            patchesRef.current.inversePatches,
        ]
        : [state, updateState]);
}
/**
 * `useMutativeReducer` is a hook that is similar to `useReducer` but it uses `mutative` to handle the state updates.
 *
 * @example
 *
 * ```ts
 * import { act, renderHook } from '@testing-library/react';
 * import { type Draft } from 'mutative';
 *
 * import { useMutativeReducer } from '../src/index';
 *
 * const { result } = renderHook(() =>
 *   useMutativeReducer(
 *     (
 *       draft: Draft<Readonly<{ count: number }>>,
 *       action: {
 *         type: 'increment';
 *       }
 *     ) => {
 *       switch (action.type) {
 *         case 'increment':
 *           draft.count += 1;
 *       }
 *     },
 *     { count: 0 }
 *   )
 * );
 * const [, dispatch] = result.current;
 * act(() => dispatch({ type: 'increment' }));
 * expect(result.current[0]).toEqual({ count: 1 });
 * ```
 */
function useMutativeReducer(
/**
 * A function that returns the next state tree, given the current state tree and the action to handle.
 */
reducer, 
/**
 * The initial state. You may optionally provide an initializer function to calculate the initial state.
 */
initializerArg, 
/**
 * An initializer function that returns the initial state. It will be called with `initializerArg`.
 */
initializer, 
/**
 * Options for the `useMutativeReducer` hook.
 */
options) {
    const patchesRef = useRef({
        patches: [],
        inversePatches: [],
    });
    //#region support strict mode and concurrent features
    const count = useRef(0);
    const renderCount = useRef(0);
    let currentCount = count.current;
    useEffect(() => {
        count.current = currentCount;
        renderCount.current = currentCount;
    });
    currentCount += 1;
    renderCount.current += 1;
    //#endregion
    const cachedReducer = useMemo(() => (state, action) => {
        const result = create(state, (draft) => reducer(draft, action), options);
        if (options === null || options === void 0 ? void 0 : options.enablePatches) {
            // check render count, support strict mode and concurrent features
            if (renderCount.current === count.current ||
                renderCount.current === count.current + 1) {
                Array.prototype.push.apply(patchesRef.current.patches, result[1]);
                // `inversePatches` should be in reverse order when multiple setState() executions
                Array.prototype.unshift.apply(patchesRef.current.inversePatches, result[2]);
            }
            return result[0];
        }
        return result;
    }, [reducer]);
    const result = useReducer(cachedReducer, initializerArg, initializer);
    useEffect(() => {
        if (options === null || options === void 0 ? void 0 : options.enablePatches) {
            // Reset `patchesRef` when the component is rendered each time
            patchesRef.current.patches = [];
            patchesRef.current.inversePatches = [];
        }
    });
    return (options === null || options === void 0 ? void 0 : options.enablePatches)
        ? [
            result[0],
            result[1],
            patchesRef.current.patches,
            patchesRef.current.inversePatches,
        ]
        : result;
}
export { useMutative, useMutativeReducer, };
